idx/memory/memory.go

```diff
@@ -0,0 +1,523 @@
+	memoryIdx := flag.NewFlagSet("memory-idx", flag.ExitOnError)
+	memoryIdx.BoolVar(&Enabled, "enabled", false, "")
+	globalconf.Register("memory-idx", memoryIdx)
+}
+
+type Tree struct {
+	Items map[string]*Node
+}
+
+type Node struct {
+	Path     string
+	Children []string
+	Leaf     bool
+}
+
+func (n *Node) String() string {
+	t := "branch"
+	if n.Leaf {
+		t = "leaf"
+	}
+	return fmt.Sprintf("%s - %s", t, n.Path)
+}
+
+// Implements the the "MetricIndex" interface
+type MemoryIdx struct {
+	sync.RWMutex
+	DefById map[string]*schema.MetricDefinition
+	Tree    map[int]*Tree
+}
+
+func New() *MemoryIdx {
+	return &MemoryIdx{
+		DefById: make(map[string]*schema.MetricDefinition),
+		Tree:    make(map[int]*Tree),
+	}
+}
+
+func (m *MemoryIdx) Init(stats met.Backend) error {
+	idxOk = stats.NewCount("idx.ok")
+	idxFail = stats.NewCount("idx.fail")
+	idxAddDuration = stats.NewTimer("idx.add_duration", 0)
+	idxGetDuration = stats.NewTimer("idx.get_duration", 0)
+	idxListDuration = stats.NewTimer("idx.list_duration", 0)
+	idxFindDuration = stats.NewTimer("idx.find_duration", 0)
+	idxDeleteDuration = stats.NewTimer("idx.delete_duration", 0)
+	return nil
+}
+
+func (m *MemoryIdx) Stop() {
+	return
+}
+
+func (m *MemoryIdx) Add(data *schema.MetricData) {
+	pre := time.Now()
+	m.Lock()
+	defer m.Unlock()
+	existing, ok := m.DefById[data.Id]
+	if ok {
+		log.Debug("metricDef with id %s already in index.", data.Id)
+		existing.LastUpdate = data.Time
+		idxOk.Inc(1)
+		idxAddDuration.Value(time.Since(pre))
+		return
+	}
+
+	def := schema.MetricDefinitionFromMetricData(data)
+	m.add(def)
+	idxAddDuration.Value(time.Since(pre))
+}
+
+func (m *MemoryIdx) Load(defs []schema.MetricDefinition) {
+	m.Lock()
+	for _, def := range defs {
+		m.add(&def)
+	}
+	m.Unlock()
+}
+
+func (m *MemoryIdx) AddDef(def *schema.MetricDefinition) {
+	pre := time.Now()
+	m.Lock()
+	defer m.Unlock()
+	if existing, ok := m.DefById[def.Id]; ok {
+		log.Debug("metricDef with id %s already in index.", def.Id)
+		existing.LastUpdate = def.LastUpdate
+		idxOk.Inc(1)
+		idxAddDuration.Value(time.Since(pre))
+		return
+	}
+	m.add(def)
+	idxAddDuration.Value(time.Since(pre))
+}
+
+func (m *MemoryIdx) add(def *schema.MetricDefinition) {
+	m.DefById[def.Id] = def
+	path := def.Name
+	//first check to see if a tree has been created for this OrgId
+	tree, ok := m.Tree[def.OrgId]
+	if !ok {
+		log.Debug("first metricDef seen for orgId %d", def.OrgId)
+		root := &Node{
+			Path:     "",
+			Children: make([]string, 0),
+			Leaf:     false,
+		}
+		m.Tree[def.OrgId] = &Tree{
+			Items: map[string]*Node{"": root},
+		}
+		tree = m.Tree[def.OrgId]
+	} else {
+		// now see if there is alread a leaf node. This happens
+		// when there are multiple metricDefs for the same path due
+		// to different tags or interval
+		if node, ok := tree.Items[path]; ok {
+			if !node.Leaf {
+				//bad data. A path cant be both a leaf and a branch.
+				log.Error(3, "Bad data, a path can not be both a leaf and a branch")
```

Comment:
this is a user error.  error logs are for errors that concern the operator of the system. We shouldn't flood our logs with errors when users send in bad keys.